net/speedtest/speedtest_test.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package speedtest
+
+import (
+	"fmt"
+	"net"
+	"testing"
+	"time"
+)
+
+func TestDownload(t *testing.T) {
+	// start up the speedtest server with a hardcoded port and address
+
+	//go StartServer("127.0.0.1", "8080")
+
+	// Create a channel to signal the server to close and defer the signal
+	// so that the server closes when the test ends.
+	killServer := make(chan bool, 1)
+	defer (func() { killServer <- true })()
+	serverPort, err := getProbablyFreePortNumber()
```

Comment:
This allocates a port, then closes it and hopes to reopen it before something else grabs that port.

Recommend instead to do the net.Listen("tcp", ":0") right here, and set serverPort from net.SplitHostPort(l.Addr().String()) here. That way you know you have that port without a race condition.

I say this because a CI machine is often running a bunch of tests, you actually can lose the race condition to own that port. Even worse, you can end up sending packets to a port which is now owned by another test, and end up causing that test to fail for inexplicable reasons.